File: backend/app/executors/slack_executor.py
from __future__ import annotations

from decimal import Decimal
from typing import Any

from app.integrations.slack_connector import SlackConnector


MOCK_SLACK_EXECUTION: dict[str, Any] = {
    "executor": "slack",
    "status": "simulated",
    "message": "Slack alert would be sent with finding and action summary.",
}


class SlackExecutor:
    """Sends or simulates Slack action alerts."""

    async def execute(
        self,
        *,
        finding: dict[str, Any],
        webhook_url: str | None = None,
        mode: str = "dry_run",
    ) -> dict[str, Any]:
        summary = self._summary(finding)
        if mode in {"simulation", "dry_run"} or not webhook_url:
            return {**MOCK_SLACK_EXECUTION, "mode": mode, "summary": summary}

        delivery = await SlackConnector(webhook_url).send_action_notification(summary)
        return {"executor": "slack", "status": "completed", "mode": mode, "summary": summary, "delivery": delivery}

    @staticmethod
    def _summary(finding: dict[str, Any]) -> str:
        savings = Decimal(str(finding.get("estimated_savings") or "0"))
        return (
            f"{finding.get('finding_type', 'optimization')} "
            f"({finding.get('severity', 'unknown')}): "
            f"estimated monthly savings {savings}. "
            f"Recommended action: {finding.get('recommended_action', 'Review manually.')}"
        )

File: backend-java/src/main/java/com/killswitch/executor/SlackExecutor.java
package com.killswitch.executor;

import com.killswitch.service.SlackService;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class SlackExecutor {
    private final SlackService slackService;

    public Map<String, Object> execute(Map<String, Object> finding, String webhookUrl, String mode) {
        String message = "KillSwitch finding: " + finding.getOrDefault("reason", finding.get("finding_type"));
        if ("simulation".equals(mode) || "dry_run".equals(mode)) {
            return Map.of("executed", false, "mode", mode, "message", message, "execution_log", "Slack alert simulated.");
        }
        Map<String, Object> delivery = slackService.sendAlert(webhookUrl, message);
        return Map.of("executed", true, "mode", mode, "delivery", delivery, "execution_log", "Slack alert sent.");
    }
}
